### DISCUSSION TITLE
How do I set up button mappings for my drawing tablet?

### DISCUSSION THREAD

**Discussion by squeaktoy** [hyprwm/Hyprland#6226] [Q&A]
On X11, I used the following commands to set keymappings for the buttons on my tablet:
```
xsetwacom set "HID 256c:006d Pad pad" button 1 key Ctrl +
xsetwacom set "HID 256c:006d Pad pad" button 2 key Ctrl -
xsetwacom set "HID 256c:006d Pad pad" button 3 key Ctrl Shift Z
xsetwacom set "HID 256c:006d Pad pad" button 8 key Ctrl Z
xsetwacom set "HID 256c:006d Pad pad" button 9 key E
xsetwacom set "HID 256c:006d Pad pad" button 10 key Alt
xsetwacom set "HID 256c:006d Pad pad" button 11 key Shift
xsetwacom set "HID 256c:006d Pad pad" button 12 key Ctrl
```

How do I do this in Hyprland?

**Comment by Agent00Ming** [comment]
Currently, Hyprland simply passes the button events to the focused client and lets the client decide what to to with it. For example, you can map what the buttons do inside of [rnote](https://github.com/flxzt/rnote) but it will only work for that application. You can use [OpenTabletDriver](https://opentabletdriver.net/) as a workaround and map your buttons inside of that.

**Reply by squeaktoy** [reply]
Programs like wev and Krita don't recognize anything when I press my tablet's buttons. I'm not sure if libinput is handling my tablet correctly. Do you know if I can manually specify which tablet model it is, like I did with the following Xorg config?:
```
Section "InputClass"
	Identifier "Huion tablets with Wacom driver"
	MatchUSBID "5543:006e|256c:006e|256c:006d"
	MatchDevicePath "/dev/input/event*"
#	MatchIsKeyboard "false"
	MatchIsTablet "true"
	Driver "wacom"
EndSection
```

**Reply by Agent00Ming** [reply]
Yeah, Krita has it's own weird method for detecting the buttons and it runs in xWayland... I don't think you can specify a model like you would in xorg. wev also does not have tablet input detection which would explain why it doesn't print anything. I suggest trying the buttons inside of rnote as a test if you are doubting that libinput is handling the buttons properly.

**Reply by squeaktoy** [reply]
rnote doesn't compile and OpenTabletDriver won't detect my tablet, so I'm not sure how to test this.

**Reply by Agent00Ming** [reply]
Well I have a CTL-6100 and the buttons work for me in rnote and such, I tried out OTD and it worked for me as well barring some driver bugs. I don't remap my buttons anymore and just use the libwacom kernel driver because it's more stable.

**Reply by squeaktoy** [reply]
Okay I just tried rnote from Flathub, and I could map 4 pad buttons, while my drawing tablet has 6 buttons. None of the buttons actually worked though.

**Reply by Agent00Ming** [reply]
that's very interesting... maybe rnote can only map 4 buttons at once? my tablet has 4 on it and I could use and map them just fine in rnote from arch extra repos...

**Comment by arximboldi** [comment]
I found a solution via `input-remapper-gtk`, but it seems somewhat buggy (the UI crashes when disabling the mapping). Also I am finding `hyprland` crashing sometimes when using the Wacom (#11641).

Any improvements or workarounds highly appreciated!

**Reply by arximboldi** [reply]
Ok, OTD seems to be working well.  No more crashing and mappings work, except the wheels of the tablet don't work yet...

I hope we get the issues via libinput.  Gnome does have such a beautiful interface to map the tablet buttons, everything working well... I'm a bit jealous :)
